File: src/tasks/cli/command/info/InfoCommand.java
package tasks.cli.command.info;

import static tasks.cli.arg.registration.Parameter.Repeatable.REPEATABLE;

import omnia.data.cache.Memoized;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import tasks.cli.arg.CliArguments;
import tasks.cli.arg.CliMode;
import tasks.cli.arg.CliUtils;
import tasks.cli.arg.registration.CommandRegistration;
import tasks.cli.arg.registration.TaskParameter;
import tasks.model.Task;

/** Canonical definition for the Info command. */
public final class InfoCommand {
  private InfoCommand() {}

  public static CommandRegistration registration(
      Memoized<CliArguments.Parser<? extends List<CliUtils.ParseResult<Task>>>> taskParser) {
    return CommandRegistration.builder()
        .cliMode(CliMode.INFO)
        .canonicalName("info")
        .aliases("i")
        .parameters(ImmutableList.of(new TaskParameter(REPEATABLE)))
        .options(ImmutableList.empty())
        .parser(() -> new InfoParser(taskParser))
        .helpDocumentation(
            "Prints all known information about a particular task, including its "
                + "description, all tasks blocking it, and all tasks it is blocking.");
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/info/InfoCommand.kt
package me.strangepan.tasks.cli.feature.info

import omnia.data.cache.Memoized
import omnia.data.cache.Memoized.Companion.memoize
import omnia.data.structure.immutable.ImmutableList
import me.strangepan.tasks.cli.command.Command
import me.strangepan.tasks.cli.command.Parameter
import me.strangepan.tasks.cli.command.TaskParameter

/** Canonical definition for the Info command.  */
object InfoCommand {
  fun registration(): Command {
    return COMMAND.value()
  }

  private val COMMAND: Memoized<Command> = memoize {
    Command.builder()
        .canonicalName("info")
        .aliases("i")
        .parameters(ImmutableList.of(TaskParameter(Parameter.Repeatable.REPEATABLE)))
        .options(ImmutableList.empty())
        .helpDocumentation("Prints all known information about a particular task, including its "
            + "description, all tasks blocking it, and all tasks it is blocking.")
  }
}